pkg/datapath/linux/node.go

```diff
@@ -1104,11 +1104,13 @@ func (n *linuxNodeHandler) nodeUpdate(oldNode, newNode *nodeTypes.Node, firstAdd
 	}
 
 	if option.Config.EnableWireguard && newNode.WireguardPubKey != "" {
-		if err := n.wgAgent.UpdatePeer(newNode.Fullname(), newNode.WireguardPubKey, newIP4, newIP6); err != nil {
-			log.WithError(err).
-				WithField(logfields.NodeName, newNode.Fullname()).
-				Warning("Failed to update wireguard configuration for peer")
-		}
+		go func() {
```

Comment:
One problem with this approach is that there is no guarantee of the order in which concurrent wgAgent.UpdatePeer will be performed.  While in practice this may be infrequent enough that mis-ordering never happens, it's possible that when two updates occur the peer may only have the first update applied last.  A more ideal solution would be to use the standard go producer-consumer model over a `chan` to ensure correct ordering.

Reply:
Thanks a lot for the PR. I share the concern about update ordering. If a peer's public key changes and we apply the update for that node out of order, then we would temporarily break the connection to that peer, since we'd assume the wrong pubkey.

Therefore I do think it's probably better to ensure that WireGuard peers are updated in the same order that the node manager observes node updates. I see two options:

- Introduce a queue that queues the updates and drains them asynchronously . This could even be hidden inside a `EnqueuePeerUpdate` function, so callers don't have to worry about channel management. The queue in this case could even be a `map` that maps `nodeName` to `{pubkey, nodeipv4, nodeipv6}` and we only keep track of the most recent event for each node.

- Move the `wgAgent` out of `linuxNodeHandler` and instead create a new `wireGuardNodeHandler` based on the [NodeHandler interface](https://github.com/cilium/cilium/blob/2c91c9a3cada1719481dadf1cef0cfedb36aa4af/pkg/datapath/types/node.go#L108). Unfortunately, the interface is overloaded and most methods would be a no-op, so option 1sounds potentially cleaner.